syntax = "proto3";

service Fileservice {

	rpc UploadFile (stream FileData) returns (ack);
	rpc DownloadFile (FileInfo) returns (stream FileData);
	rpc FileSearch (FileInfo) returns (ack);
	rpc ReplicateFile (stream FileData) returns (ack);
	rpc FileList (UserInfo) returns (FileListResponse);
	rpc FileDelete (FileInfo) returns (ack);
	rpc UpdateFile (stream FileData) returns (ack);

	rpc getStatus (HeartBeatRequest) returns (HeartBeatResponse) {}
}

message FileData {
	string username = 1;
	string filename =2;
	bytes data =3;
}

message ack {
	bool success =1;
	string message = 2;
}

message UserInfo {
	string username=1;
}

message FileInfo {
	string username = 1;
	string filename = 2;
}

message FileListResponse {
	string Filenames =1;
}

// The request message containing ip address and leader status
message HeartBeatRequest {
  string ip = 1;
  bool leader = 2;
}

// The response message containing the status of the server
message HeartBeatResponse {
  string ip = 1;
  bool live = 2;
  float cpu_usage = 3;
  float disk_space = 4;
  int32 num_process = 5;
  int32 num_thread = 6;
  float idle = 7;
  float tot_mem = 8;
  float used_mem = 9;
  float data_read_per_sec = 10;
  float data_write_per_sec = 11;
  float data_recieve_per_sec = 12;
  float data_sent_per_sec = 13;
}
